### PR TITLE
Feature/gaussiandensity values

### LINKED ISSUE #666: Gaussian convolution of the parameter
Hi,

I think that might be useful to extend GuassianDensity class to allow the convolution of the parameter. FX, if some parameter p is assigned to every atom in the box, its density function could be written as convolution of parameter for atom *l* with Dirac's delta function:

image

For some tasks (e.g. studying phase transitions) the blurring of the parameter density function might be needed ([see for example](https://elifesciences.org/articles/13150)). Adding such option to GuassianDensity class is straightforward and would have solved the issue. The parameter list might be passed as an argument to *compute* method and C++ code at the very last step, the gaussian distribution should be multiplied by the parameter for the current point.

### PR DESCRIPTION
**Pull request by pbuslaev**
This a pull request for issue #666. I have implemented the convolution of the parameter array `p` with Gaussians. The changes concern 

- `freud/density.pyx`: Added the parameter values=None to compute method. The type of the values is float

-  `freud/_density.pxd`: Added an argument for the array of values

- `cpp/density/GaussianDensity.h`: Added argument matching the type of that added to _density.pxd

- `cpp/density/GaussianDensity.cc`: Added the argument and used the input values to calculate the convolution

### CODE REVIEW

**Comment by vyasr** [comment]
@pbuslaev thanks for the contribution! I'll look at your code as soon as I can.

Currently your code is failing our style checks, which we do using `flake8`. If you `pip install flake8` you'll be able to run `flake8 freud` from the root of the repo and it will tell you how to address those issues. For future commits, you can set up a pre-commit hook, which [we've documented in our development guide](https://freud.readthedocs.io/en/stable/reference/development/howtoadd.html). We do plan to switch to using the [`pre-commit`](https://pre-commit.com/) framework for this at some point, but for now the old style flake8 hook will get the job done. Thanks!

**Comment by pbuslaev** [comment]
@vyasr Thanks for your comments! I have updated the style.

File: freud/density.pyx
Comment (bdice) [inline freud/density.pyx]: Mention that the "Gaussian blur" is equivalent to every point having a value of 1.
```suggestion
                convolution. Calculates Gaussian blur (equivalent to providing
                a value of 1 for every point) if :code:`None`. (Default value
                = :code:`None`).
```

File: freud/density.pyx
Comment (bdice) [inline freud/density.pyx]: 1. We can avoid an unnecessary double-float conversion here. We should specify the dtype for `np.ones` since it defaults to `float64`, but we don't need to specify the dtype for `_convert_array` since it defaults to `float32`. 
2. We don't need to check the shape in `_convert_array` since we already know its shape is correct (by construction). We only have to check the shape of a user-provided array of values.

```suggestion
                np.ones(nq.points.shape[0], dtype=np.float32))
```

File: freud/density.pyx
Comment (bdice) [inline freud/density.pyx]: ~This can be marked as `const`.~ _Edit:_ I don't think this can be marked as `const` if you use the code above to support `NULL`. (Cython requires variable declaration to be outside of a conditional, but you have to assign the variable conditionally so you can't make it `const`.)
```suggestion
        cdef const float[::1] l_values = values
```

**Review by bdice — CHANGES_REQUESTED** [review]
Overall this PR is very good! Nice work, @pbuslaev. I made a few suggestions that you can apply directly.

Please add tests for using a set of distinct values. This is the file that contains tests: https://github.com/glotzerlab/freud/blob/master/tests/test_density_GaussianDensity.py

One minor improvement I would like to see is allowing `NULL` for the values array in C++. (We use `NULL` instead of `nullptr` for compatibility with Cython, though `nullptr` is typically preferred.) We are considering making our C++ API "public" in the future so that other libraries could use freud internally, so having the "None" behavior match in Python and C++ would be nice. You can see an example of how to do that here:
https://github.com/glotzerlab/freud/blob/deef8b5678e10c1a8157e4b8e617f80aea2f1706/freud/box.pyx#L344-L385
https://github.com/glotzerlab/freud/blob/deef8b5678e10c1a8157e4b8e617f80aea2f1706/cpp/box/Box.h#L364-L390

File: freud/density.pyx
Comment (pbuslaev) [reply freud/density.pyx]: @bdice thanks a lot for your comments. I have added tests and `nullptr` to the latest commit

Comment (pbuslaev) [reply freud/density.pyx]: @bdice Thanks for your valuable comments. I have committed your suggestions about Gaussian blur and double-float conversion, added tests and nullptr

File: cpp/density/GaussianDensity.cc
Comment (vyasr) [inline cpp/density/GaussianDensity.cc]:
```suggestion
void GaussianDensity::compute(const freud::locality::NeighborQuery* nq, const float* values)
```

File: cpp/density/GaussianDensity.h
Comment (vyasr) [inline cpp/density/GaussianDensity.h]:
```suggestion
    void compute(const freud::locality::NeighborQuery* nq, const float* values = NULL);
```

File: freud/_density.pxd
Comment (vyasr) [inline freud/_density.pxd]:
```suggestion
                     const float*) except +
```

File: tests/test_density_GaussianDensity.py
Comment (vyasr) [inline tests/test_density_GaussianDensity.py]:
```suggestion
        gd.compute(system, values)
```

This line is the reason your tests are currently failing, you're running into test errors because you're providing system as a kwarg but not values.

File: tests/test_density_GaussianDensity.py
Comment (vyasr) [inline tests/test_density_GaussianDensity.py]:
```suggestion
        gd.compute(system, values)
```

Same as above.

File: tests/test_density_GaussianDensity.py
Comment (vyasr) [inline tests/test_density_GaussianDensity.py]:
```suggestion
        system = freud.data.make_random_system(
```

Since you're just handing the system around without using the box or points independently, you can just pass the tuple around.

File: tests/test_density_GaussianDensity.py
Comment (vyasr) [inline tests/test_density_GaussianDensity.py]:
```suggestion
        system = freud.data.make_random_system(
```

**Review by vyasr — CHANGES_REQUESTED** [review]
This is looking really good! I've left some comments that you should be able to directly apply. One additional request: can you update the docstring for `GaussianDensity` to include the formula (just the simple product-sum from #666)? It's not a complicated formula, but it's always helpful for new users to be able to associate the code with the underlying math without having to think too hard.

**Comment by vyasr** [comment]
@pbuslaev I wasn't sure if you had enabled our modification of your branch. You did, so I went ahead and applied my proposed edits and merged in master for you. You can still see all my comments (even though they're marked as resolved) if you're curious what I changed. Only outstanding request I have is to update the documentation. Specifically, above the `Args` section of the `GaussianDensity` docstring if you could add an integral (sum) showing what the GaussianDensity actually does, and how the default convolution kernel is the indicator function but can be modified using the `values` argument, that would be fantastic!

File: cpp/density/GaussianDensity.cc
Comment (bdice) [inline cpp/density/GaussianDensity.cc]:
```suggestion
            const float value = (values != NULL) ? values[idx] : 1.0f;
```

File: freud/density.pyx
Comment (bdice) [inline freud/density.pyx]: We can leave off the `dtype=np.float32` since it's currently the default.
```suggestion
                values, shape=(nq.points.shape[0], ))
```

File: tests/test_density_GaussianDensity.py
Comment (bdice) [inline tests/test_density_GaussianDensity.py]: I think this only works because `num_points = 1`. I think the total should be `num_points * np.mean(values)`, otherwise written as `np.sum(values)`. We should update the tests to `assert np.isclose(np.sum(gd.density), num_points, atol=1e-4)` above as well.
```suggestion
        assert np.isclose(np.sum(gd.density), np.sum(values), atol=1e-4)
```

File: tests/test_density_GaussianDensity.py
Comment (bdice) [inline tests/test_density_GaussianDensity.py]:
```suggestion
        assert np.isclose(np.sum(gd.density), np.sum(values), atol=1e-4)
```

File: freud/density.pyx
Comment (bdice) [inline freud/density.pyx]: Added vector notation -- I'm going to go ahead and commit this.
```suggestion
        p(\vec{r}) = \sum_i \frac{1}{2\pi \sigma^2}
        \exp \left(-\frac{(\vec{r}-\vec{r}_i)^2}{2\sigma^2}\right) p_i
```

**Review by bdice — APPROVED** [review]
Really great work @pbuslaev, thanks for the PR! I made some final updates to make the tests a little more rigorous, and I approve.

**Comment by bdice** [comment]
@pbuslaev One more request: please complete the PR checklist below (you might have deleted this checklist when creating the PR message). The primary things are indicating that you agree to the CONTRIBUTING document and updating the changelog and credits to include your contributions.

## Checklist:

- [ ] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] I have updated the documentation (if relevant).
- [x] I have added tests that cover my changes (if relevant).
- [ ] All new and existing tests passed.
- [ ] I have updated the [credits](https://github.com/glotzerlab/freud/blob/master/doc/source/reference/credits.rst).
- [ ] I have updated the [Changelog](https://github.com/glotzerlab/freud/blob/master/ChangeLog.md).

**Comment by bdice** [comment]
@pbuslaev Don't worry about the failing CI checks from ReadTheDocs, I will rebase this PR on `master` and fix that before merging.

**Comment by pbuslaev** [comment]
> @pbuslaev One more request: please complete the PR checklist below (you might have deleted this checklist when creating the PR message). The primary things are indicating that you agree to the CONTRIBUTING document and updating the changelog and credits to include your contributions.

For some reason, I can't select anything in the checklist. However, I did everything according to it

## Checklist: 

 - [x]  I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.
 - [x]  My code follows the code style of this project.
 - [x]  I have updated the documentation (if relevant).
 - [x]  I have added tests that cover my changes (if relevant).
 - [x]  All new and existing tests passed.
 - [x]  I have updated the [credits](https://github.com/glotzerlab/freud/blob/master/doc/source/reference/credits.rst). 
 - [x]  I have updated the [Changelog](https://github.com/glotzerlab/freud/blob/master/ChangeLog.md).

**Comment by bdice** [comment]
@pbuslaev Fantastic, thanks! I'll get this updated and merged (we changed our build system to use scikit-build while you were working on this PR). I appreciate the great work!

**Comment by bdice** [comment]
This can be merged when tests pass.

**Pull request merged**
